classes/helper/mahara_portfolio.php

```diff
@@ -0,0 +1,273 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Manage connection to Mahara for portfolio 'HTML Lite' export.
+ *
+ * @package   plagiarism_urkund
+ * @author    David Balch <david.balch@catalyst-eu.net>
+ * @copyright 2021 onwards Catalyst IT Europe <http://catalyst-eu.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace plagiarism_urkund\helper;
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Class mahara_portfolio
+ *
+ * @package   plagiarism_urkund
+ * @copyright 2021 onwards Catalyst IT Europe <http://catalyst-eu.net>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class mahara_portfolio {
+
+    private $instanceid;
+    private $context;
+    private $userid;
+    private $relateduserid;
+    private $submission;
+    private $fileinfo;
+    private $host;
+    private $oauthurl;
+    private $oauthkey;
+    private $oauthsecret;
+    private $wstoken;
+    private $fileurl;
+    private $zipfile;
+
+    /**
+     * Constructor.
+     *
+     * @param array[int] Data about this assignsubmission_maharaws submission:
+     *      $instanceid instance id of the assignment.
+     *      $submissionid id of the submission.
+     *      $userid id of the user submitting.
+     *      $relateduserid id of the user being submitted onbehalf of.
+     * @throws dml_exception
+     */
+    public function __construct($data) {
+        global $CFG, $DB;
+        $this->instanceid = $data->instanceid;
+        $this->userid = $data->userid;
+        $this->relateduserid = $data->relateduserid;
+
+        // Get submission context id from context instance id.
+        $this->context = \context_module::instance($this->instanceid);
+
+        // Load submission details.
+        $this->submission = $this->get_mahara_submission($data->submissionid);
+
+        // Set site/assignment config.
+        $plagiarismvalues = $DB->get_records_menu('plagiarism_urkund_config',
+                array('cm' => $this->instanceid), '', 'name, value');
+        $this->host = $CFG->wwwroot;
+        $this->oauthurl = $plagiarismvalues['urkund_maharawsurl'];
+        $this->oauthkey = $plagiarismvalues['urkund_maharawshtmllitekey'];
+        $this->oauthsecret = $plagiarismvalues['urkund_maharawshtmllitesecret'];
+        $this->wstoken = $plagiarismvalues['urkund_maharawshtmllitetoken'];
+
+        // Check if there's a zip file already downloaded by previous processing.
+        $fs = get_file_storage();
+        $fileinfo = array(
```

Comment:
I feel like we be setting userid in this array - but I haven't checked to see if it's setting the correct userid later on in the code to the student (owner of the file.)

also - as this is new user data we are storing within the moodle site we'll need to implement privacy api functions to allow retrieiving and deleting of these files. (hopefully just the individual content files, as we're deleting the zip files after using them right?)